### PR TITLE
Fixes/4781 treeview setting bound selected item null

### LINKED ISSUE #4780: TreeView SelectSingleItem causing StackOverFlow
Hey,

The SelectSingleItem in TreeView clearing SelectedItems causing SelectedItemCollectionChanged which cause an update to the SelectedItem and becasue SelectedItems is empty the value that updates is null. https://github.com/AvaloniaUI/Avalonia/blob/2a89ba9b24732a17dc0418e09842fa377580ffa8/src/Avalonia.Controls/TreeView.cs#L220

To repro - 
**Control**
`  
<TreeView Items="{Binding Items}" SelectedItem="{Binding SelectedItem}"/>
`

**ViewModel property**
`
public MyItem SelectedItem
{
    get => _selectedItem;
    set
    {
        if (Equals(value, _selectedItem) || value == null) return;
        _selectedItem = value;
        OnPropertyChanged(nameof(SelectedItem));
    }
}
`
Removing the check for null fixes the issue, but that shouldn't happen if there aren't any null items in the Items to begin with.

In this screenshot you can see that PointerPressed (on an item) caused SelectSingleItem which caused SelectedItems.Clear() which caused the SelectedItem to be update with null.

image

### PR DESCRIPTION
**Pull request by danwalmsley**
## What does the pull request do?

Fixes the behavior of SelectedItem property when user changes the selection on treeview.

## What is the current behavior?

The users bound property will go:
1) has a selected value.
2) user selects another item with the mouse.
3) property is updated to null.
4) property is updated to new value.

This is because currently the method `SelectSingleItem` calls `Clear` on `SelectedItems`. This sends out INCC events.
It then calls `Add`.

Also currently it is assumed `SelectedItem` can only be set if it is already null. I removed this check as we can switch from 1 item to the next.

## What is the updated/expected behavior with this PR?

We supress the raising of the events caused by the `Clear` call, using an existing flag (`_syncingSelectedItems`).

We allow `SelectedItem` to be set even if its not already `null`.

User will now get the binding updated as expected.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues

Fixes #4780

### CODE REVIEW

File: src/Avalonia.Controls/TreeView.cs [hunk 1]
```diff
@@ -353,7 +355,7 @@ private void SelectedItemsAdded(IList items)
                 MarkItemSelected(item, true);
             }
 
-            if (SelectedItem == null && !_syncingSelectedItems)
```
Comment (grokys) [inline src/Avalonia.Controls/TreeView.cs:356]: I don't think removing this check is correct. Elsewhere, if you have a multi-selecting control and select one item followed by another then `SelectedItem` remains as the first selected item, i.e. `SelectedItem` represents the first selected item in a multi-selection and not the last. I think this is the case in WPF as well.

**Review by grokys — CHANGES_REQUESTED** [review]

File: src/Avalonia.Controls/TreeView.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Controls/TreeView.cs:356]: I think the solution would be to surround the whole of `SelectSingleItem` in a `_syncingSelectedItems` fence and manually `SetAndRaise` the `SelectedItem` property.

Comment (grokys) [reply src/Avalonia.Controls/TreeView.cs:356]: Which would also mean the `_syncingSelectedItems` fence in `SelectedItem.set` is unnecessary.

File: tests/Avalonia.Controls.UnitTests/TreeViewTests.cs [hunk 2]
```diff
@@ -1288,5 +1325,27 @@ public bool Match(object data)
         private class DerivedTreeView : TreeView
         {
         }
+
+        private class TestDataContext : ReactiveObject
+        {
+            private int _counter;
```
Comment (grokys) [inline tests/Avalonia.Controls.UnitTests/TreeViewTests.cs:1331]: `_counter` is unused.

File: src/Avalonia.Controls/TreeView.cs [hunk 3]
```diff
@@ -219,8 +217,13 @@ private void SubscribeToSelectedItems()
 
         private void SelectSingleItem(object item)
         {
-            SelectedItems.Clear();
+            var oldValue = SelectedItem;
```
Comment (grokys) [inline src/Avalonia.Controls/TreeView.cs:220]: `oldValue` doesn't seem to be used?

**Review by grokys — CHANGES_REQUESTED** [review]

File: tests/Avalonia.Controls.UnitTests/TreeViewTests.cs (hunk 2 above)
Comment (danwalmsley) [reply tests/Avalonia.Controls.UnitTests/TreeViewTests.cs]: removed.

File: src/Avalonia.Controls/TreeView.cs (hunk 3 above)
Comment (danwalmsley) [reply src/Avalonia.Controls/TreeView.cs]: removed.

**Review by grokys — APPROVED** [review]

**Pull request merged**